server/cluster/unsafe_recovery_controller.go

```diff
@@ -121,452 +142,629 @@ func (u *unsafeRecoveryController) RemoveFailedStores(failedStores map[uint64]st
+	newestRegionTree, peersMap := u.buildUpFromReports()
+	hasPlan := false
+
+	findForceLeader := func(peersMap map[uint64][]*regionItem, region *metapb.Region) *regionItem {
+		var leader *regionItem
+		for _, peer := range peersMap[region.GetId()] {
+			if peer.report.IsForceLeader {
+				leader = peer
+				break
 			}
-			recoveredRanges.ReplaceOrInsert(regionItem{newRegion})
 		}
-		if len(creates) != 0 || update != nil {
-			storeRecoveryPlan, exists := u.storeRecoveryPlans[storeID]
-			if !exists {
-				u.storeRecoveryPlans[storeID] = &pdpb.RecoveryPlan{}
-				storeRecoveryPlan = u.storeRecoveryPlans[storeID]
-			}
-			storeRecoveryPlan.Creates = append(storeRecoveryPlan.Creates, creates...)
-			if update != nil {
-				storeRecoveryPlan.Updates = append(storeRecoveryPlan.Updates, update)
+		return leader
+	}
+
+	// Check the regions in newest Region Tree to see if it can still elect leader
+	// considering the failed stores
+	newestRegionTree.tree.Ascend(func(item btree.Item) bool {
+		region := item.(*regionItem).Region()
+		if !u.canElectLeader(region) {
+			leader := findForceLeader(peersMap, region)
+			if leader == nil {
+				// can't find the force leader, skip
+				return true
 			}
-		} else if _, healthy := healthyRegions[region.Id]; !healthy {
-			// If this peer contributes nothing to the recovered ranges, and it does not belong to a healthy region, delete it.
-			storeRecoveryPlan, exists := u.storeRecoveryPlans[storeID]
-			if !exists {
-				u.storeRecoveryPlans[storeID] = &pdpb.RecoveryPlan{}
-				storeRecoveryPlan = u.storeRecoveryPlans[storeID]
+			storeRecoveryPlan := u.getRecoveryPlan(leader.storeID)
+			storeRecoveryPlan.Demotes = append(storeRecoveryPlan.Demotes,
+				&pdpb.DemoteFailedVoters{
+					RegionId:     region.GetId(),
+					FailedVoters: u.getFailedPeers(leader.Region()),
+				},
+			)
+			hasPlan = true
+		}
+		return true
+	})
+
+	// Tombstone the peers of region not presented in the newest region tree
+	for storeID, storeReport := range u.storeReports {
+		for _, peerReport := range storeReport.PeerReports {
+			regionID := peerReport.GetRegionState().Region.Id
+			if !newestRegionTree.contains(regionID) {
+				if u.canElectLeader(peerReport.GetRegionState().Region) {
+					log.Warn("find invalid peer but it has quorum")
+				} else {
+					// the peer is not in the valid regions, should be deleted directly
+					storeRecoveryPlan := u.getRecoveryPlan(storeID)
+					storeRecoveryPlan.Tombstones = append(storeRecoveryPlan.Tombstones, regionID)
+				}
 			}
-			storeRecoveryPlan.Deletes = append(storeRecoveryPlan.Deletes, region.Id)
 		}
 	}
-	// There may be ranges that are covered by no one. Find these empty ranges, create new regions that cover them and evenly distribute newly created regions among all stores.
+	return hasPlan
+}
+
+func (u *unsafeRecoveryController) generateCreateEmptyRegionPlan() bool {
+	newestRegionTree, _ := u.buildUpFromReports()
```

Comment:
How about using a flag to control if we need to build the peer map?

Reply:
I believe it would be optimized by the compiler somehow?